tests/h/search/core_test.py

```diff
@@ -0,0 +1,243 @@
+        annotation_ids.
+        """
+        annotation = Annotation(shared=True)
+        reply_1 = Annotation(shared=True, references=[annotation.id])
+        reply_2 = Annotation(shared=True, references=[annotation.id])
+
+        result = search.Search(pyramid_request).run({})
+
+        assert result.annotation_ids == matchers.UnorderedList([annotation.id, reply_1.id,
+                                                                reply_2.id])
+
+    def test_replies_that_dont_match_the_search_arent_included(self, factories, pyramid_request,
+                                                               Annotation):
+        """Replies that don't match the search query aren't included.
+
+        Not even if the top-level annotations that they're replies to _are_
+        included.
+        """
+        user = factories.User()
+        reply_user = factories.User()
+        annotation = Annotation(userid=user.userid, shared=True)
+        reply = Annotation(userid=reply_user.userid, references=[annotation.id], shared=True)
+
+        result = search.Search(pyramid_request).run(
+            # Search for annotations from ``user``, so that ``reply_user``'s
+            # reply doesn't match.
+            params={"user": user.userid},
+        )
+
+        assert reply.id not in result.annotation_ids
+
+    def test_replies_from_different_pages_arent_included(self, pyramid_request, Annotation):
+        """Replies may not be on the same page of results as their annotations."""
+        # First create an annotation and a reply.
+        annotation = Annotation(shared=True)
+        reply = Annotation(references=[annotation.id], shared=True)
+
+        # Now create 19 more annotations so that the original annotation is
+        # pushed onto the second page of the search results, but the reply is
+        # still on the first page.
+        for _ in range(19):
+            Annotation(shared=True)
+
+        # The reply is one the first page of search results, but the original annotation isn't.
+        result = search.Search(pyramid_request).run(params={"offset": 0, "limit": 20})
+        assert reply.id in result.annotation_ids
+        assert annotation.id not in result.annotation_ids
+
+        # The original annotation is on the second page of search results, but the reply isn't.
+        result = search.Search(pyramid_request).run(params={"offset": 20})
+        assert reply.id not in result.annotation_ids
+        assert annotation.id in result.annotation_ids
+
+    def test_replies_can_come_before_annotations(self, pyramid_request, Annotation):
+        """A reply may appear before its annotation in the search results.
+
+        Things are returned in updated order so normally a reply would appear
+        before the annotation that it is a reply to in the search results.
+        """
+        annotation = Annotation(shared=True)
+        reply = Annotation(references=[annotation.id], shared=True)
```

Comment:
Is it **guaranteed** that the reply will have a later updated date here? If not we could have a flakey test.